### PR TITLE
Document Sensu Enterprise Console API

### LINKED ISSUE #561: Document Sensu Enterprise Console API
## This is an issue with:

- [ ] Bug (site functionality or styling)
- [ ] Errata (fix needed for doc content)
- [x] New content (guide wanted)
- [ ] Update (Add missing or refresh existing content)

## Expected Behavior
A reference doc that describes the SE API

## Current Behavior
Currently not present.

## Possible Solution
Write it.

### PR DESCRIPTION
**Pull request by apaskulin**
Create API docs for the Sensu Enterprise Console API.

- [Helpscout ticket](https://secure.helpscout.net/conversation/515983639/9054/?folderId=1211663)
- #561, #313 

## Description

Based on #561, the endpoints we need to document are as follows:

Comparison of list above with existing docs in sensu-core/api

| Console endpoint(s) | Content copied from Sensu Core docs? |
| --- | --- |
/aggregates (GET) | ✅ 
/aggregates/:aggregate (GET, DELETE) |✅ 
/checks (GET) | ✅ 
/checks/:check (GET, DELETE) | Added DELETE
/checks/:check (GET, DELETE) | duplicate
/clients (GET) | ✅ 
/clients/:client (GET, DELETE) | ✅ 
/clients/:client/history (GET) | ✅ 
/config (GET) | Added
/datacenters (GET) | Added
/datacenters/:name (GET) | Added
/events (GET) | ✅ 
/events/:client/:check (DELETE) | ✅ 
/request (POST) | ✅ 
/results/:client/:check (DELETE) | ✅ 
/silenced (GET, POST) | ✅ 
/silenced/clear (POST) | ✅ 
/stashes/:path (DELETE) | ✅ 
/stashes (GET, POST) | ✅ 
/subscriptions (GET) | Added
/subscriptions/:subscription (GET) | Added

Notes:

POST operations require dc attribute to specify the target datacenter.

When accessing an individual resource, Uchiwa will try to find it amongst the datacenters but in case there's multiple resources with the same name, it will return all resources found with the HTTP code Multiple Choices (300). In order to avoid that, it's possible to use a query string that specifies the datacenter (e.g. ?dc=foo).

### Testing checklist

- [x] /aggregates (GET)
- [x] /aggregates/:aggregate (GET)
- [x] /aggregates/:aggregate (DELETE)
- [x] /checks (GET)
- [x] /checks/:check (GET)
- [ ] ~/checks/:check (DELETE) - Not supported~
- [x] /request (POST)
- [x] /clients (GET)
- [x] /clients/:client (GET)
- [x] /clients/:client (DELETE)
- [ ] ~/clients/:client/history (GET) - Deprecated, no docs~
- [x] /config (GET)
- [x] /datacenters (GET)
- [x] /datacenters/:name (GET)
- [x] /events (GET)
- [x] /events/:client/:check (DELETE)
- [x] /results/:client/:check (DELETE)
- [x] /silenced (GET)
- [x] /silenced (POST)
- [x] /silenced/clear (POST) - Removed option to clear via subscription+check (instead of by ID) due to 404s
- [x] /stashes/:path (DELETE) 
- [x] /stashes (GET)
- [x] /stashes (POST)
- [x] /subscriptions (GET)
- [ ] ~/subscriptions/:subscription (GET)~ - Not supported

### To do

- [x] Copy and paste available endpoints from Core API docs
- [x] Find-and-replace 4567 with 3000
- [x] Locate source info for endpoints without existing docs
- [x] Document endpoints without existing docs
- [x] Proofread and test
- [x] Add `dc` query parameter and 300 response
- [x] Create API overview content
- [ ] Resolve outstanding items in the testing list
- [x] Update messaging to reflect capabilities of Console API
  - http://localhost:1313/sensu-enterprise-dashboard/2.12/#what-is-the-sensu-enterprise-console
  - http://localhost:1313/sensu-enterprise-dashboard/2.12/rbac/overview/#rbac-for-the-sensu-enterprise-console-api
- [x] Update https://docs.sensu.io/sensu-enterprise-dashboard/2.12/rbac/overview/#methods-attributes
- [ ] Copy to all versions post-review

## Motivation and Context

Closes #561

### CODE REVIEW

**Comment by cwjohnston** [comment]
> /datacenters (GET)

* Implemented in Uchiwa at https://github.com/sensu/uchiwa/blob/1.3.1-1/uchiwa/server.go#L625 and https://github.com/sensu/uchiwa/blob/1.3.1-1/uchiwa/datacenter.go
* Returns an array reflecting the results of the /info API for each configured datacenter 

> /datacenters/:name (GET)

* Implemented in Uchiwa at https://github.com/sensu/uchiwa/blob/1.3.1-1/uchiwa/server.go#L584
* Returns /info API result for a named datacenter

> /config (GET)

* Implemented in Uchiwa at https://github.com/sensu/uchiwa/blob/1.3.1-1/uchiwa/server.go#L553 and https://github.com/sensu/uchiwa/blob/1.3.1-1/uchiwa/config/config.go#L317
* Returns redacted Uchiwa/Sensu Enterprise Dashboard configuration hash

> /subscriptions (GET)

* Implemented in Uchiwa at https://github.com/sensu/uchiwa/blob/1.3.1-1/uchiwa/server.go#L1277 and https://github.com/sensu/uchiwa/blob/1.3.1-1/uchiwa/filters/filters.go#L78
* Returns an array of hashes describing subscription/datacenter pairs, filtered on subscriptions allowed by requester's role

> /subscriptions/:subscription (GET)

* Implemented in Uchiwa at https://github.com/sensu/uchiwa/blob/1.3.1-1/uchiwa/server.go#L1247
* Unclear to me what should be returned here... requesting valid or invalid subscriptions both returns HTTP 200

**Comment by apaskulin** [comment]
👋 @palourde Would you mind looking over this PR to document the Enterprise Console API?

Also, I wasn't able to get a response from the /subscriptions/:subscription (GET) endpoint. Should this endpoint be working in the Console API?

**Comment by palourde** [comment]
@apaskulin I just realized the `/subscriptions/:subscription` endpoint is not even used by the dashboard and it has a really weird behaviour; it simply tells you if the provided subscription would be allowed based on the configured RBAC roles, not if the subscription exists or not. We could either remove it from the documentation or open an issue, not sure if the API users would actually need that.

File: content/sensu-enterprise-dashboard/2.12/api/aggregates.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+]
+{{< /highlight >}}
+
+#### API specification {#aggregates-get-specification}
+
+/aggregates (GET) | 
+------------------|------
+description       | Returns the list of named aggregates by `name` and datacenter (`dc`)
+example url       | http://hostname:3000/aggregates
+parameters        | <ul><li>`max_age`:<ul><li>**required**: false</li><li>**type**: Integer</li><li>**description**: the maximum age of results to include, in seconds.</li></ul></li></ul>
```
Comment (palourde) [inline content/sensu-enterprise-dashboard/2.12/api/aggregates.md]: The console API does not support the `max_age` parameter

File: content/sensu-enterprise-dashboard/2.12/api/aggregates.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+}
+{{< /highlight >}}
+
+#### API specification {#aggregatesname-get-specification}
+
+/aggregates/:name (GET) | 
+------------------------|------
+description             | Returns the list of aggregates for a given check.
+example url             | http://hostname:3000/aggregates/elasticsearch
+parameters              | <ul><li>`max_age`:<ul><li>**required**: false</li><li>**type**: Integer</li><li>**description**: the maximum age of results to include, in seconds.</li></ul></li></ul><ul><li>`dc`:<ul><li>**required**: false</li><li>**type**: String</li><li>**description**: If the aggregate name is present in multiple datacenters, specifying the `dc` parameter will return only the aggregate found in that datacenter.</li><li>**example**: `http://hostname:3000/aggregates/elasticsearch?dc=us_west1`</li></ul></li></ul>
```
Comment (palourde) [inline content/sensu-enterprise-dashboard/2.12/api/aggregates.md]: `max_age` is also not supported

File: content/sensu-enterprise-dashboard/2.12/api/clients.md [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    "environment": "development",
+    "timestamp": 1458625739,
+    "version": "1.5.0"
+  }
+]
+{{< /highlight >}}
+
+_NOTE: for larger Sensu installations it may be undesirable to get the entire
+[client registry][1] in a single API request. The `/clients` API provides
+pagination controls via the [`limit` and `offset` url parameters][7] (see
```
Comment (palourde) [inline content/sensu-enterprise-dashboard/2.12/api/clients.md]: Pagination is not available in the Console API

File: content/sensu-enterprise-dashboard/2.12/api/clients.md [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+below)._
+
+#### API Specification {#clients-get-specification}
+
+/clients (GET) | 
+---------------|------
+description    | Returns a list of clients by `name` and datacenter (`dc`). Since clients use the Sensu Core package, the `version` returned by this endpoint reflects the current version of Sensu Core instead of Sensu Enterprise.
+example url    | http://hostname:3000/clients
+parameters     | <ul><li>`limit`<ul><li>**required**: false</li><li>**type**: Integer</li><li>**description**: The number of clients to return.</li><li>**example**: `http://hostname:3000/clients?limit=100`</li></ul></li><li>`offset`<ul><li>**required**: false</li><li>**type**: Integer</li><li>**depends**: `limit`</li><li>**description**: The number of clients to offset before returning items.</li><li>**example**: `http://hostname:3000/clients?limit=100&offset=100`</li></ul></li></ul>
```
Comment (palourde) [inline content/sensu-enterprise-dashboard/2.12/api/clients.md]: idem

File: content/sensu-enterprise-dashboard/2.12/api/clients.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+parameters                | <ul><li>`invalidate`<ul><li>**required**: false</li><li>**type**: Boolean</li><li>**description**: If the Sensu client should be invalidated, disallowing further client keepalives and check results until the client is successfully removed from the client registry.</li><li>**example**: `http://hostname:3000/clients/i-424242?invalidate=true`</li></ul><li>`invalidate_expire`<ul><li>**required**: false</li><li>**type**: Integer</li><li>**description**: If the Sensu client should be invalidated for a specified amount of time (in seconds), disallowing further client keepalives and check results even after the client is successfully removed from the client registry.</li><li>**example**: `http://hostname:3000/clients/i-424242?invalidate=true&invalidate_expire=3600`</li></ul></li></ul><ul><li>`dc`:<ul><li>**required**: false</li><li>**type**: String</li><li>**description**: If the client name is present in multiple datacenters, specifying the `dc` parameter will access only the client found in that datacenter.</li><li>**example**: `http://hostname:3000/clients/i-424242?dc=us_west1`</li></ul></li></ul>
```
Comment (palourde) [inline content/sensu-enterprise-dashboard/2.12/api/clients.md]: `invalidate` & `invalidate_expire` are not supported unfortunately

File: content/sensu-enterprise-dashboard/2.12/api/config.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  - [`/config` (GET)](#config-get)
+
+## The `/config` API endpoint {#the-config-endpoint}
+
+### `/config` (GET)
+
+The `/config` API provides HTTP GET access to the Sensu Enterprise
+configuration. Sensitive setting values are redacted by
+default, unless the URL query parameter `redacted` is set to `false`,
+e.g. `/config?redacted=false.`
```
Comment (palourde) [inline content/sensu-enterprise-dashboard/2.12/api/config.md]: The `redacted` parameter is not supported (the config will be redacted)

File: content/sensu-enterprise-dashboard/2.12/api/config.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+}
+{{< /highlight >}}
+
+#### API Specification {#config-get-specification}
+
+/config (GET) | 
+----------------|------
+description     | Returns the Sensu Enterprise configuration.
+example url     | http://hostname:3000/config
+parameters      | <ul><li>`redacted`:<ul><li>**required**: false</li><li>**type**: Boolean</li><li>**description**: If sensitive setting values should be redacted.</li><li>**default**: true</li></ul>
```
Comment (palourde) [inline content/sensu-enterprise-dashboard/2.12/api/config.md]: Not supported in the Console API

File: content/sensu-enterprise-dashboard/2.12/api/silenced.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+]
+{{< /highlight >}}
+
+#### API specification {#silenced-get-specification}
+
+/silenced (GET) | 
+----------------|------
+description     | Returns a list of silence entries.
+example url     | http://hostname:3000/silenced
+parameters      | <ul><li>`limit`:<ul><li>**required**: false</li><li>**type**: Integer</li><li>**description**: The number of silence entries to return.</li></ul><li>`offset`:<ul><li>**required**: false</li><li>**type**: Integer</li><li>**depends**: `limit`</li><li>**description**: The number of silence entries to offset before returning items.</li></ul></li></ul>
```
Comment (palourde) [inline content/sensu-enterprise-dashboard/2.12/api/silenced.md]: Both parameters are not supported

File: content/sensu-enterprise-dashboard/2.12/api/stashes.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+]
+{{< /highlight >}}
+
+#### API specification {#stashes-get-specification}  
+
+/stashes (GET) | 
+---------------|------
+description    | Returns a list of stashes by `path` and datacenter (`dc`).
+example url    | http://hostname:3000/stashes
+parameters     | <ul><li>`limit`:<ul><li>**required**: false</li><li>**type**: Integer</li><li>**description**: The number of stashes to return.</li></ul></li><li>`offset`:<ul><li>**required**: false</li><li>**type**: Integer</li><li>**depends**: `limit`</li><li>**description**: The number of stashes to offset before returning items.</li></ul></li></ul>
```
Comment (palourde) [inline content/sensu-enterprise-dashboard/2.12/api/stashes.md]: idem

**Review by palourde — CHANGES_REQUESTED** [review]
Awesome job!

Unfortunately the Console API does not support any query parameter except `dc`.

Also, there's a `/health` endpoint (see https://docs.uchiwa.io/uchiwa/1.0/api/health/), not sure if it's worth adding.

**Comment by apaskulin** [comment]
You're the best, @palourde! I've removed the unsupported parameters, omitted /subscriptions/:subscription, and added /health.

**Review by palourde — APPROVED** [review]

**Pull request merged**